gr-digital/lib/header_buffer.cc

```diff
@@ -78,81 +78,34 @@ void header_buffer::add_field64(uint64_t data, int len, bool bs)
+template <class T>
+T header_buffer::extract_field(int pos, int len, bool bs, bool lsb_first)
 {
-    if (len > 8) {
-        throw std::runtime_error("header_buffer::extract_field for "
-                                 "uint8_t length must be <= 8");
+    if (len > 8 * (int)sizeof(T)) {
+        throw std::runtime_error(
+            std::string("header_buffer::extract_field for length must be <= ") +
+            std::to_string(8 * sizeof(T)));
     }
 
-    uint8_t field = 0x00;
+    T field = 0;
     std::vector<bool>::iterator itr;
-    for (itr = d_input.begin() + pos; itr != d_input.begin() + pos + len; itr++) {
-        field = (field << 1) | ((*itr) & 0x1);
-    }
-
-    return field;
-}
-
-uint16_t header_buffer::extract_field16(int pos, int len, bool bs)
-{
-    if (len > 16) {
-        throw std::runtime_error("header_buffer::extract_field for "
-                                 "uint16_t length must be <= 16");
-    }
-
-    uint16_t field = 0x0000;
-    std::vector<bool>::iterator itr;
-    for (itr = d_input.begin() + pos; itr != d_input.begin() + pos + len; itr++) {
-        field = (field << 1) | ((*itr) & 0x1);
-    }
-
-    if (bs) {
-        volk_16u_byteswap(&field, 1);
-    }
-
-    return field;
-}
-
-uint32_t header_buffer::extract_field32(int pos, int len, bool bs)
-{
-    if (len > 32) {
-        throw std::runtime_error("header_buffer::extract_field for "
-                                 "uint32_t length must be <= 32");
-    }
-
-    uint32_t field = 0x00000000;
-    std::vector<bool>::iterator itr;
-    for (itr = d_input.begin() + pos; itr != d_input.begin() + pos + len; itr++) {
-        field = (field << 1) | ((*itr) & 0x1);
-    }
-
-    if (bs) {
-        volk_32u_byteswap(&field, 1);
+    if (lsb_first) {
+        for (itr = d_input.begin() + pos + len - 1; itr >= d_input.begin() + pos; itr--) {
+            field = (field << 1) | ((*itr) & 0x1);
+        }
+    } else {
+        for (itr = d_input.begin() + pos; itr != d_input.begin() + pos + len; itr++) {
+            field = (field << 1) | ((*itr) & 0x1);
+        }
     }
 
     return field;
 }
 
-uint64_t header_buffer::extract_field64(int pos, int len, bool bs)
-{
-    if (len > 64) {
-        throw std::runtime_error("header_buffer::extract_field for "
-                                 "uint64_t length must be <= 64");
-    }
-
-    uint64_t field = 0x0000000000000000;
-    std::vector<bool>::iterator itr;
-    for (itr = d_input.begin() + pos; itr != d_input.begin() + pos + len; itr++) {
-        field = (field << 1) | ((*itr) & 0x1);
-    }
-
-    if (bs) {
-        volk_64u_byteswap(&field, 1);
-    }
-
-    return field;
-}
+template uint8_t header_buffer::extract_field(int, int, bool, bool);
+template uint16_t header_buffer::extract_field(int, int, bool, bool);
+template uint32_t header_buffer::extract_field(int, int, bool, bool);
+template uint64_t header_buffer::extract_field(int, int, bool, bool);
```

Comment:
```suggestion
template DIGITAL_API uint8_t header_buffer::extract_field(int, int, bool, bool);
template DIGITAL_API uint16_t header_buffer::extract_field(int, int, bool, bool);
template DIGITAL_API uint32_t header_buffer::extract_field(int, int, bool, bool);
template DIGITAL_API uint64_t header_buffer::extract_field(int, int, bool, bool);
```
According to the comments on https://stackoverflow.com/questions/51347113/dll-export-class-template-member-func-unresolved-external-symbol-any-chan, which seems to be the same case, MSVC won't be happy until you add the export declaration to these explicit instantiations. I hope.